### ISSUE TITLE
collection validation

### ISSUE THREAD

**Issue by etombi** [JeremySkinner/FluentValidation#952]
### System Details

- FluentValidation version: 8.0.0
- Web Framework version (asp.net 4.6.1):

### Issue Description

Hi i have a following scenario:

the validators are registered using this code

:public class ValidationRegistry : Registry
{
    public ValidationRegistry()
    {
        AssemblyScanner.FindValidatorsInAssemblyContaining<ValidationRegistry>()
          .ForEach(result =>
          {
              For(result.InterfaceType)
                 .Use(result.ValidatorType);
          });
    }
}

We have already validator for class Person
now when i creating the validator for List<Person> its looks like it conflicts with validator for Person and Validate for list is not triggering.. instead of it triggers the single Person validator
any suggestion how to workaround this behavior ?

**Comment by JeremySkinner** [comment]
Not sure I'm totally understanding....are you saying both validators for the same type? If so then that's not going to work if you're trying to register them both as `IValidator<Person>` with the container. If you want to register with your container as `IValidator<Person>` then you'll need to make sure you only have a single validator for the type. You can break up the different scenarious using [rule sets](https://fluentvalidation.net/start#rulesets)

**Comment by etombi** [comment]
yes i guest its initilated here? AssemblyScanner.FindValidatorsInAssemblyContaining<TValidator>()
                .ForEach(result =>
                {
                    // add concrete validators only
                    if (!result.ValidatorType.IsAbstract) {
                        For(result.InterfaceType)
                            .Singleton()
                            .Use(result.ValidatorType);
                    }
                });

the problem is that a part of validation it is DB queries so i want to aggregate them into a single call .

**Comment by JeremySkinner** [comment]
That code registers validator types with your IoC container, my question was specifically where are you *executing* the validator?

**Comment by etombi** [comment]
i guess its automatically..
only lines of code in project are :
FluentValidationConfigurer.ConfigureFluentValidation(); in global asax and 
 FluentValidationModelValidatorProvider.Configure(provider =>
            {
                provider.ValidatorFactory = new StructureMapValidatorFactory();
            });
            ValidatorOptions.CascadeMode = CascadeMode.StopOnFirstFailure;
in config cs file

**Comment by etombi** [comment]
i have two  validators:
`public class PeopleValidator : AbstractValidator<People>`
and
`public class PeopleListValidator : AbstractValidator<List<People>>`
and only pepole validatior is executed when sending a list to a controllers method

**Comment by JeremySkinner** [comment]
Yes, I understand what you're saying. 

Unfortunately you can't do this with the automatic validation. MVC does not invoke FluentValidation if the root-level object is a collection. It will only invoke FluentValidation for the individual elements within the collection. This is not a limitation of FluentValidation, but rather MVC's own validation infrastructure will not invoke FluentValidation for collection parameters. 

This does work with newer version of ASP.NET (ASP.NET Core 2.0 and 2.1), but with legacy MVC5 this isn't something that can be done I'm afraid. You'd need to resort to manually invoking the validator if you wanted to do this.

**Comment by etombi** [comment]
if we are wrapping the list of person in object like 

```
PersonWrapper {
public List<Person> people { get; set; }
}
```
the validator of person and the validator of PersonWrapper  executed both

**Comment by JeremySkinner** [comment]
Yes; I would expect both to be executed in that case. The issue is only when the root-level type is a collection.

**Comment by etombi** [comment]
How can we prevent the second execution ?

**Comment by JeremySkinner** [comment]
The only way to do that would be to not register the `PeopleValidator` with the container (ie you'd have to exclude it from the call to `AssemblyScanner.FindValidatorsInAssemblyContaining`). Unfortunately this means that the `PeopleValidator` would *never* execute automatically, even when your controller action takes a `Person` instance directly, and you would have to resort to manual validation. 

This is unfortunately one of the many limitations of MVC's validation infrastructure.

**Comment by etombi** [comment]
Thanks a lot for your help  !!!

**Comment by JeremySkinner** [comment]
One workaround I just thought of would be to utilise rulesets within your `PeopleValidator`. 

If you wrap all your rules inside a custom ruleset, then they won't be executed as part of the List action. For any actions where you do want the person validator be executed directly, you could then use the `CustomizeValidatorAttribute` to opt-in to the rulesets ([like this example](https://fluentvalidation.net/aspnet#validator-customization))

It's messy but it might solve it for you.

**Comment by etombi** [comment]
i dont see the boolean (SKIP) argument for CustomValidatorAttribute its interceptor =type,properties=string,ruleset=string

**Comment by JeremySkinner** [comment]
`Skip` is only available in ASP.NET Core. You'd need to use the ruleset approach I mentioned above.

**Comment by etombi** [comment]
if i understand you right in this case only the person validator  will be executed ,but i need the opposite - to execute the list Validator and skip the PersonValidator

**Comment by JeremySkinner** [comment]
Yes, that's what it would do (execute the list validator and skip the PersonValidator).

1) Wrap *all* rules inside the PersonValidator in a `RuleSet`. 
2) For actions where you *do* need the `PersonValidator` to execute, use `[CustomizeValidator(RuleSet="the ruleset you just created")]
3) For the List action do *not* include the `CustomizeValidator`. This will ensure the rules are not executed.

**Comment by etombi** [comment]
i thought something like : 
```
 private class ListValidatorInterceptor : IValidatorInterceptor
        {
            public ValidationResult AfterMvcValidation(ControllerContext controllerContext, ValidationContext validationContext, ValidationResult result)
            {
                return result;
            }

            public ValidationContext BeforeMvcValidation(ControllerContext controllerContext, ValidationContext validationContext)
            {
                validationContext.RootContextData.Add("skip", "true");
                return validationContext;
            }
        }

```
and in the validator : 

 ```
public override FluentValidation.Results.ValidationResult Validate(ValidationContext<People> context)
        {
            if (context.RootContextData.Keys.Contains("skip"))
            {
                return new FluentValidation.Results.ValidationResult();
            }
            return base.Validate(context);
        }

```

and pass this one  in a method i need to validate only list ?

**Comment by JeremySkinner** [comment]
Yes, that approach should work and is probably cleaner than the ruleset approach I suggested.

**Comment by etombi** [comment]
Thanks a lot for your help !!!
